src/ZkFold/Symbolic/Algorithms/Hash/SHA2.hs

```diff
@@ -0,0 +1,368 @@
+   , ToWords (ByteString (ChunkSize algorithm) element) (ByteString (WordSize algorithm) element)
+   , Concat (ByteString (WordSize algorithm) element) (ByteString (8 * WordSize algorithm) element)
+   , ToWords (ByteString (PaddedLength k (ChunkSize algorithm) (2 * WordSize algorithm)) element) (ByteString (ChunkSize algorithm) element)
+   )
+
+-- | A generalised version of SHA2. It is agnostic of the ByteString base field.
+-- Sample usage:
+--
+-- >>> bs = fromConstant (42 :: Natural) :: ByteString 8 (Zp BLS12_381_Scalar)
+-- >>> hash = sha2 @"SHA256" bs
+--
+sha2
+    :: forall (algorithm :: Symbol) element k
+    .  SHA2 algorithm element k
+    => ByteString k element -> ByteString (ResultSize algorithm) element
+sha2 messageBits = sha2Blocks @algorithm @element chunks
+    where
+        paddedMessage :: ByteString (PaddedLength k (ChunkSize algorithm) (2 * WordSize algorithm)) element
+        paddedMessage = sha2Pad @(ChunkSize algorithm) @(2 * WordSize algorithm) messageBits
+
+        chunks :: [ByteString (ChunkSize algorithm) element]
+        chunks = toWords paddedMessage
+
+-- | Pad the input bytestring according to the rules described in @PaddedLength@
+sha2Pad
+    :: forall (padTo :: Natural) (lenBits :: Natural) (k :: Natural) element
+    .  KnownNat k
+    => KnownNat (PaddedLength k padTo lenBits)
+    => Finite element
+    => FromConstant Natural element
+    => ShiftBits (ByteString (PaddedLength k padTo lenBits) element)
+    => AdditiveSemigroup (ByteString (PaddedLength k padTo lenBits) element)
+    => Extend (ByteString k element) (ByteString (PaddedLength k padTo lenBits) element)
+    => ByteString k element -> ByteString (PaddedLength k padTo lenBits) element
+sha2Pad bs = grown + fromConstant padValue
+    where
+        l :: Natural
+        l = natVal $ Proxy @k
+
+        diff :: Natural
+        diff = (natVal $ Proxy @(PaddedLength k padTo lenBits)) -! (natVal $ Proxy @k)
+
+        padValue :: Natural
+        padValue = 2 P.^ (diff -! 1) P.+ l
+
+        grown :: ByteString (PaddedLength k padTo lenBits) element
+        grown = extend bs `shiftBitsL` diff
+
+
+-- | This allows us to calculate hash of a bytestring represented by a Natural number.
+-- This is only useful for testing when the length of the test string is unknown at compile time.
+-- This should not be exposed to users (and they probably won't find it useful anyway).
+--
+instance (KnownNat n, Finite a, FromConstant Natural a) => ToWords Natural (ByteString n a) where
+    toWords = P.reverse . toWords'
+        where
+            toWords' :: Natural -> [ByteString n a]
+            toWords' 0 = []
+            toWords' n = fromConstant (n `mod` base) : toWords' (n `div` base)
+
+            base :: Natural
+            base = 2 P.^ (value @n)
+
+
+-- | Constraints required for a SHA2 of a Natural number.
+--
+type SHA2N algorithm element =
```

Comment:
@vlasin, and this is the reason why I once suggested to introduce a `class Symbolic a` with instances in `Zp` and `ArithmeticCircuit`